File: ai_analyst/api/models/reflect.py
"""Reflect aggregation API response models (PR-REFLECT-1, PR-REFLECT-3)."""

from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel, Field

from ai_analyst.api.models.ops import ResponseMeta

ArtifactState = Literal["present", "missing", "malformed"]


class ScanBounds(BaseModel):
    max_runs: int
    inspected_dirs: int
    valid_runs: int
    skipped_runs: int


class SuggestionEvidence(BaseModel):
    metric_name: Literal["override_rate", "no_trade_rate"]
    metric_value: float
    threshold: float
    sample_size: int


class Suggestion(BaseModel):
    rule_id: Literal["OVERRIDE_FREQ_HIGH", "NO_TRADE_CONCENTRATION"]
    severity: Literal["warning"] = "warning"
    category: Literal["persona", "pattern"]
    target: str
    message: str
    evidence: SuggestionEvidence


class PersonaStats(BaseModel):
    persona: str
    participation_count: int
    skip_count: int
    fail_count: int
    participation_rate: float
    override_count: int
    override_rate: Optional[float] = None
    stance_alignment: Optional[float] = None
    avg_confidence: Optional[float] = None
    flagged: bool = False
    navigable_entity_id: Optional[str] = None


class PersonaPerformanceResponse(ResponseMeta):
    threshold: int = 10
    threshold_met: bool
    scan_bounds: ScanBounds
    stats: list[PersonaStats]
    suggestions: list[Suggestion] = Field(default_factory=list)


class VerdictCount(BaseModel):
    verdict: str
    count: int


class PatternBucket(BaseModel):
    instrument: str
    session: str
    run_count: int
    threshold_met: bool
    verdict_distribution: list[VerdictCount]
    no_trade_rate: Optional[float] = None
    flagged: bool = False


class PatternSummaryResponse(ResponseMeta):
    threshold: int = 10
    scan_bounds: ScanBounds
    buckets: list[PatternBucket]
    suggestions: list[Suggestion] = Field(default_factory=list)


class ArtifactStatus(BaseModel):
    run_record: ArtifactState
    usage_jsonl: ArtifactState
    usage_json: ArtifactState


class RunBundleResponse(ResponseMeta):
    run_id: str
    artifact_status: ArtifactStatus
    run_record: dict = Field(default_factory=dict)
    usage_summary: Optional[dict] = None
    usage_jsonl: list[dict] = Field(default_factory=list)

File: ui/src/shared/api/reflect.ts
// ---------------------------------------------------------------------------
// Reflect endpoint clients — shapes from PR_REFLECT_2_SPEC.md.
// Three read-only endpoints consuming PR-REFLECT-1 backend.
// ---------------------------------------------------------------------------

import { apiFetch, type ApiResult } from "./client";
import type { ResponseMeta } from "./ops";

// ---- Persona Performance types ----

export type PersonaStats = {
  persona: string;
  participation_count: number;
  skip_count: number;
  fail_count: number;
  participation_rate: number;
  override_count: number;
  override_rate: number | null;
  stance_alignment: number | null;
  avg_confidence: number | null;
  flagged: boolean;
};

export type ScanBounds = {
  max_runs: number;
  inspected_dirs: number;
  valid_runs: number;
  skipped_runs: number;
};

export type PersonaPerformanceResponse = ResponseMeta & {
  threshold: number;
  threshold_met: boolean;
  scan_bounds: ScanBounds;
  stats: PersonaStats[];
};

// ---- Pattern Summary types ----

export type VerdictCount = {
  verdict: string;
  count: number;
};

export type PatternBucket = {
  instrument: string;
  session: string;
  run_count: number;
  threshold_met: boolean;
  verdict_distribution: VerdictCount[];
  no_trade_rate: number | null;
  flagged: boolean;
};

export type PatternSummaryResponse = ResponseMeta & {
  threshold: number;
  scan_bounds: ScanBounds;
  buckets: PatternBucket[];
};

// ---- Run Bundle types ----

export type ArtifactState = "present" | "missing" | "malformed";

export type ArtifactStatus = {
  run_record: ArtifactState;
  usage_jsonl: ArtifactState;
  usage_json: ArtifactState;
};

export type RunBundleResponse = ResponseMeta & {
  run_id: string;
  artifact_status: ArtifactStatus;
  run_record: Record<string, unknown>;
  usage_summary: Record<string, unknown> | null;
  usage_jsonl: Record<string, unknown>[];
};

// ---- Fetch params ----

export type FetchPersonaPerformanceParams = {
  maxRuns?: number;
};

export type FetchPatternSummaryParams = {
  maxRuns?: number;
};

// ---- Endpoint functions ----

export function fetchPersonaPerformance(
  params: FetchPersonaPerformanceParams = {},
): Promise<ApiResult<PersonaPerformanceResponse>> {
  const searchParams = new URLSearchParams();
  if (params.maxRuns != null)
    searchParams.set("max_runs", String(params.maxRuns));
  const query = searchParams.toString();
  const path = query
    ? `/reflect/persona-performance?${query}`
    : "/reflect/persona-performance";
  return apiFetch<PersonaPerformanceResponse>(path);
}

export function fetchPatternSummary(
  params: FetchPatternSummaryParams = {},
): Promise<ApiResult<PatternSummaryResponse>> {
  const searchParams = new URLSearchParams();
  if (params.maxRuns != null)
    searchParams.set("max_runs", String(params.maxRuns));
  const query = searchParams.toString();
  const path = query
    ? `/reflect/pattern-summary?${query}`
    : "/reflect/pattern-summary";
  return apiFetch<PatternSummaryResponse>(path);
}

export function fetchRunBundle(
  runId: string,
): Promise<ApiResult<RunBundleResponse>> {
  return apiFetch<RunBundleResponse>(`/reflect/run/${encodeURIComponent(runId)}`);
}
